addons/binding/org.openhab.binding.max/src/main/java/org/openhab/binding/max/internal/MaxConsoleCommandExtension.java

```diff
@@ -0,0 +1,137 @@
+
+    private ThingRegistry thingRegistry;
+
+    public MaxConsoleCommandExtension() {
+        super("max", "Additional EQ3 MAX! commands.");
+    }
+
+    @Override
+    public void execute(String[] args, Console console) {
+        if (args.length > 0) {
+            switch (args[0]) {
+                case SUBCMD_BACKUP:
+                    for (Thing thing : findDevices(CUBEBRIDGE_THING_TYPE)) {
+                        MaxCubeBridgeHandler handler = getHandler(thing.getUID().toString());
+                        if (handler != null) {
+                            handler.backup();
+                            console.println("Creating backup for " + thing.getUID().toString());
+                        }
+                    }
+                    break;
+                case SUBCMD_REBOOT:
+                    if (args.length > 1) {
+                        MaxCubeBridgeHandler handler = getHandler(args[1]);
+                        if (handler != null) {
+                            handler.cubeReboot();
+                        } else {
+                            console.println("Could not find cube " + args[1]);
+                            printMaxDevices(console, SUPPORTED_BRIDGE_THING_TYPES_UIDS);
+                        }
+                    } else {
+                        console.println("Specify cube to reboot.");
+                        printMaxDevices(console, SUPPORTED_BRIDGE_THING_TYPES_UIDS);
+                    }
+                    break;
+                default:
+                    console.println("Unknown MAX! sub command '" + args[1] + "'");
+                    printUsage(console);
+                    break;
+            }
+        } else {
+            printUsage(console);
+            printMaxDevices(console, SUPPORTED_THING_TYPES_UIDS);
+        }
+    }
+
+    private List<Thing> findDevices(ThingTypeUID type) {
+        List<Thing> devs = new ArrayList<Thing>();
+        for (Thing thing : thingRegistry.getAll()) {
+            if (thing.getThingTypeUID().equals(type)) {
+                devs.add(thing);
+            }
+        }
+        return devs;
+    }
+
+    private MaxCubeBridgeHandler getHandler(String thingId) {
+        MaxCubeBridgeHandler handler = null;
+        try {
+            ThingUID bridgeUID = new ThingUID(thingId);
+            Thing thing = thingRegistry.get(bridgeUID);
+            if ((thing != null) && (thing.getHandler() != null)
+                    && (thing.getHandler() instanceof MaxCubeBridgeHandler)) {
+                handler = (MaxCubeBridgeHandler) thing.getHandler();
+            }
+        } catch (Exception e) {
+            handler = null;
+        }
+        return handler;
+    }
+
+    private void printMaxDevices(Console console, Set<ThingTypeUID> deviceTypes) {
+        console.println("Known MAX! devices: ");
+        for (ThingTypeUID type : deviceTypes) {
+            for (Thing thing : findDevices(type)) {
```

Comment:
The `getAll` in your method `findDevices` is a relative expensive operation as it makes a copy of Things. So you might rewrite it a bit to pass the set of devices to `findDevices` and put the `for (ThingTypeUID type : deviceTypes) {` inside the `getAll` loop. Otherwise it will make copies for all devicesTypes passed in the set.

Reply:
yes, makes sense as well